Add support for running fuzzing jobs in Kubernetes

NishantBansal2003 opened the pull request:
Fixes: #33 

This PR adds support for running fuzzing in Kubernetes. Now, the user needs to specify the `fuzz.in-cluster` boolean flag if they want to run fuzzing in K8s mode. Also, the user can specify `fuzz.namespace` to indicate which K8s namespace to use (default is the `default` namespace).

This PR also ensures that we now have two modes of fuzzing: one is local using Docker, and the other is using this K8s mode.

## Workflow
The workflow is similar to the container setup, where instead of creating and starting a fuzzing container, we now create a fuzzing job in the given namespace. This fuzzing job requires 2 GB of memory and 1 CPU. The job uses a fixed service account name: `go-continuous-fuzz-sa`, and a fixed PVC named `go-continuous-fuzz-pvc`. There are some guidelines on this K8s setup, which are mentioned in the `docs/USAGE.md` of this PR.
Once the job is created, it is set with a backoff limit of 0 so that if there is a fuzz crash, we do not restart the pod. (This means if the pod stopped due to a crash other than a fuzz crash, we will gracefully stop the application so that we can debug changes.) After that, we wait for the pod to be either in a terminal or running state so that we can stream its logs. If there is a fuzz crash, we handle that gracefully - otherwise, we wait until the job is completed or failed. If it fails, we return the error and stop the application - if the context is cancelled, we simply delete the job.

## Additional Information
I also added the Helm chart, which is pretty much standard, so the user can use it as is. For the templates of pod, configmap, and PVC, I did not provide their templates because I thought they are pretty much user-dependent (though I have referenced the YAML definitions used during e2e tests in the documentation, which can serve as templates). Instead, I provided general guidelines for creating pod, configmap, and PVC for the go-continuous-fuzz project. Let me know if the templates need to be provided?

At last, I tested this code with Minikube, so we need to test this code with AWS as well, and based on that, we need to provide templates for pod, configmap, and PVC.

djkazic commented:
We can stand up a ec2 box and then y'all can run kind in it if that works. Can't really do a proper cluster at the moment.

NishantBansal2003 commented:
> We can stand up a ec2 box and then y'all can run kind in it if that works.

OK, but does this completely simulate AWS behavior, or is it more like Minikube?



> Can't really do a proper cluster at the moment.

No worries. we can currently test using Kind or Minikube, review and finalise the code, and later once we have cluster access run the actual tests and merge?

djkazic commented:
@NishantBansal2003 I wouldn't count on real cluster access. Kind is like minikube.

djkazic commented on go-continuous-fuzz-chart/templates/serviceaccount.yaml:
```diff
@@ -0,0 +1,8 @@
+{{- if .Values.serviceAccount.create -}}
```
We can probably make this unconditionally create the svcaccount as the pod being spawned always relies on it existing.

djkazic commented on go-continuous-fuzz-chart/templates/rolebinding.yaml:
```diff
@@ -0,0 +1,15 @@
+{{- if .Values.rbac.create -}}
```
This should be unconditional to ensure the svcaccount has the correct rolebinding

NishantBansal2003 commented:
Rebased and addressed the review comments. Also added support for github issue verification using k8s, works perfectly with the same container logic.

Ref: https://github.com/NishantBansal2003/go-continuous-fuzz/pull/19 for CI run results

cc: @gijswijs

gijswijs commented on go-continuous-fuzz-chart/templates/configmap.yaml:
```diff
@@ -0,0 +1,26 @@
+
+    [Project]
+    project.workspace-path = "{{ .Values.project.workspacePath }}"
+    project.src-repo = "https://oauth2:{{ .Values.github.authToken}}@{{ trimPrefix "https://" .Values.project.srcRepo }}"
```
The GitHub PAT is rendered into a ConfigMap here (and again on line 17 for the crash-repo). ConfigMaps aren't encrypted at rest and are readable by anyone with `get configmap`. The AWS creds already go in a Secret (`awssecret.yaml`), so this looks like an oversight — suggest moving `github.authToken` into a Secret and consuming it from env rather than interpolating it into the config URL here.

NishantBansal2003 replied:
Done in https://github.com/go-continuous-fuzz/go-continuous-fuzz/pull/44/commits/c1f929a18fe0ca54e0274779325542820d24ac1f

I standardized this approach across both docker and k8s to ensure that secrets are never written to disk and are instead consumed at runtime.

gijswijs commented on cluster.go:
```diff
@@ -0,0 +1,302 @@
+			BackoffLimit: ptr.To(int32(0)),
+			Template: corev1.PodTemplateSpec{
+				Spec: corev1.PodSpec{
+					ServiceAccountName: "go-continuous-fuzz-sa",
```
The spawned fuzz Job runs as `go-continuous-fuzz-sa` but doesn't set `AutomountServiceAccountToken: false` on the PodSpec, so the SA's K8s API token is mounted into the fuzz workload — which runs third-party target code under mutation. Recommend `AutomountServiceAccountToken: ptr.To(false)` here; the fuzz Job itself doesn't call the K8s API (only the controller does).

NishantBansal2003 replied:
Done in https://github.com/go-continuous-fuzz/go-continuous-fuzz/pull/44/commits/c1f929a18fe0ca54e0274779325542820d24ac1f
